### ISSUE TITLE
Training and validation accuracy and loss from result and graph

### ISSUE THREAD

**Issue by John12Reaper** [ultralytics/yolov5#1246]
## ❔Question
Sir, 
1. I wanted to know which are the training accuracy and validation accuracy 
and also training loss and validation loss in the results.txt?

2. How can I plot the training and validation accuracy in a single graph 
and training and validation loss in another graph?

Here is the result

![issue](https://user-images.githubusercontent.com/45218568/97683815-b48c4000-1ac2-11eb-9ef2-1ad1133d553e.PNG)

Please it would be really helpful.
Thank You

**Comment by glenn-jocher** [comment]
@John12Reaper accuracy is a classification metric. YOLOv5 does object detection. The metrics available are displayed to screen, saved to results.txt, plotted to results.png after training completion and logged to Tensorboard. 

Suggest you start with the custom training tutorial, where all of this is explained.
https://docs.ultralytics.com/yolov5/tutorials/train_custom_data

**Comment by glenn-jocher** [comment]
@John12Reaper this explains the difference:

<img width="1563" alt="Screenshot 2020-10-27 at 14 20 53" src="https://user-images.githubusercontent.com/26833433/97689991-ac210e80-1a9c-11eb-9309-d99f071d17b7.png">

**Comment by John12Reaper** [comment]
@glenn-jocher Thank You Sir
So this means there is no training accuracy or validation accuracy in results.txt 
Then what is the mAP@ : .5 and mAP@ .5 : .95 ? 
And Sir What does all the matrix mean in the results.txt kindly if you tell or suggest any?
Sorry Sir I am new to object detection.

**Comment by dongjuns** [comment]
Yes, you are right.
There is no training accuracy or validation accuracy metric, but an mAP metric on your validation dataset.
And I think you need some detailed information about mAP,
<https://jonathan-hui.medium.com/map-mean-average-precision-for-object-detection-45c121a31173>

**Comment by John12Reaper** [comment]
@dongjuns Thank You, Sir
I really appreciate your kind help. 
Sir, please could you tell me how to print of val AP and train AP in a single graph
or val Loss and train Loss in a single graph? 
It will be quite helpful.

**Comment by dongjuns** [comment]
@John12Reaper You're welcome, look into these functions.
https://github.com/ultralytics/yolov5/blob/c8c5ef36c9a19c7843993ee8d51aebb685467eca/utils/general.py#L1263
https://github.com/ultralytics/yolov5/blob/c8c5ef36c9a19c7843993ee8d51aebb685467eca/utils/general.py#L1240

And you can draw training loss and validation loss in a single graph like this.
![results_overlay](https://user-images.githubusercontent.com/23523114/97778174-acd9a380-1b75-11eb-91c7-9801e863d099.png)

Move your results.txt file into your YOLOv5 directory, I'm using docker and in my case, YOLOv5 directory path is /usr/src/app.
Then you can get your results.png with this script.
```
root@21e424cd396c:/usr/src/app# ls
Dockerfile   detect.py   requirements.txt  road_detect.py  train.py
LICENSE      hubconf.py    runs     yolov5x.pt
README.md    inference     sotabench.py     tutorial.ipynb
__pycache__  merge.py     test.py       utils
data         models      results.txt    weights

root@21e424cd396c:/usr/src/app# python
Python 3.6.10 |Anaconda, Inc.| (default, May  8 2020, 02:54:21) 
[GCC 7.3.0] on linux
Type "help", "copyright", "credits" or "license" for more information.
>>> from utils.general import *
>>> plot_results_overlay()

root@21e424cd396c:/usr/src/app# ls
Dockerfile   detect.py    requirements.txt    results.txt    weights
LICENSE      hubconf.py    road_detect.py    train.py       
README.md    inference     runs    yolov5x.pt
__pycache__  merge.py      sotabench.py    tutorial.ipynb
data         models      results.png       test.py        utils
```

**Comment by John12Reaper** [comment]
@dongjuns Sir, You are Great.
I have got the results.png as given below:
![results](https://user-images.githubusercontent.com/45218568/97783585-04eac700-1bc3-11eb-8dd2-bec2b2ef0019.png)

Sir, Thank You again.
I have also seen the functions and made some changes to get other relational graphs.
but Sir, One last favor, could you please help?
Is there any way to make these graphs in Tensorboard? just the above graphs.
Cause in TensorBoard it is quite smooth for the evaluation of the performance.

**Comment by dongjuns** [comment]
@John12Reaper Don't say 'Sir' please!!! :sob: 
Anyway, I don't use tensorboard but maybe you could get an idea from this link and train.py file.
<https://github.com/ultralytics/yolov5/blob/master/tutorial.ipynb>

Good luck-!

**Comment by glenn-jocher** [comment]
@John12Reaper tensorboard automatically logs all trainings. Tensorboard Instructions are displayed at the beginning of every training run:

```bash
Start Tensorboard with "tensorboard --logdir runs/", view at http://localhost:6006/
```
<img width="1361" alt="Screenshot 2020-10-31 at 19 33 31" src="https://user-images.githubusercontent.com/26833433/97787051-02cb3d00-1bb0-11eb-912b-ef31b25c99a5.png">

**Comment by ZwNSW** [comment]
@dongjuns I watched the conversation between you and@John12Reaper, but I don’t think I understand what you said, what should I do to get the images of val and train. Or how to modify the code to get orange lines and blue lines? Mine only has blue lines, and no validation set lines.
![image](https://user-images.githubusercontent.com/57338462/97959058-8328b280-1de9-11eb-8bd1-5f8b000f3a5f.png)

**Comment by carlosmunozledesma** [comment]
Hi, where can I find the **plot_results_overlay()** function in the last version of github? Thanks a lot.

**Comment by glenn-jocher** [comment]
@carlosmunozledesma `plots_results_overlay()` has been removed from latest code as we did not think there was significant community adoption, and the switch to CSV logging required new maintenance for the results plotting functions.

One feature that many people don't know about though is that you can still overlay multiple results by placing multiple results.csv files in a single directory `dir` and then:
```python
from utils.plots import *; plot_results(dir='path/to/dir')
```

**Comment by HarshThakur121** [comment]
Sir i got yolov3_training_last.weights/yolov3.weights after training then how could i plot validation and accuracy graph in yolov3?
![ss](https://github.com/ultralytics/yolov5/assets/85275733/2fa6e261-7085-46c1-910e-afbc26139f50)

**Comment by glenn-jocher** [comment]
@thakurgaurisingh hello!

YOLOv5 repository does not support YOLOv3 anymore. The `yolov3_training_last.weights/yolov3.weights` files are the trained weights for YOLOv3.

To plot validation and accuracy graph for YOLOv3, you can try using a visualization tool such as TensorBoard. You can log your training and validation results to a `logdir` and then run TensorBoard to visualize the training process. Here is an example command:

```
tensorboard --logdir=path/to/logdir
```

Please note that you need to have TensorFlow installed to use TensorBoard.

I hope this helps! Let me know if you have any further questions.

**Comment by mahtabfaraji1** [comment]
Hi everyone! I just checked the conversation, and it doesn't seem like there is a solution for calculating mAp for the training phase. How should I modify the code to include mAp calculations in the training? (Even it doesn't make sense, I need that part)

Thank you in advance for your help.

**Comment by HarshThakur121** [comment]
Yes please modify the code so that we can calculate in training phase only

**Comment by glenn-jocher** [comment]
Hello @thakurgaurisingh Faraji and @thakurgaurisingh,

Calculating mAP during the training phase is not a common practice because it can be computationally expensive and slow down the training process. However, if you need to calculate mAP for each training batch for some specific reason, you would need to modify the training loop to include mAP calculation after each batch or epoch.

This would involve using the validation dataset or a subset of the training data to compute the mAP. You would need to run the predictions of the model on this dataset and compare them to the ground truth annotations to calculate the mAP.

Please note that this is an advanced modification and requires a good understanding of the codebase. It's also important to remember that doing this could significantly increase your training time.

If you're not familiar with how to implement this, it might be a good idea to consult with someone who has experience in deep learning and PyTorch, or consider hiring a consultant to assist you with the modifications.

Remember to always backup your code before making such changes, and test your modifications thoroughly to ensure they work as expected.

Good luck with your project! 🚀
